lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,338 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ThreadPoolExecutor.h"
+
+#include <queue>
+
+namespace glow {
+namespace runtime {
+
+ExecutionState::ExecutionState(RunIdentifierTy id, const DAGNode *root,
```

Comment:
Hm. I initially wrote this constructor in order to create all required `Contexts` beforehand and to insert the required `Placeholders` into them without backing `Tensors` to avoid synchronization and locking later. I quickly realized `Context` doesn't support that, but left this as is. Can we afford to take the time to do all this setup beforehand? Or should per-node state only be created when it is needed?

Reply:
Also, `Context` doesn't have tracking for the number of node parents that are done, or these little intermediate input `Contexts`.